libtiledbsoma/src/reindexer/reindexer.cc

```diff
@@ -0,0 +1,163 @@
+/**
+ * @file   reindexer.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the a pybind11 api in the SOMA C++ library.
+ */
+
+#include "reindexer.h"
+#include <thread_pool/thread_pool.h>
+#include <unistd.h>
+#include <thread>
+#include "khash.h"
+#include "soma/enums.h"
+#include "soma/soma_array.h"
+#include "utils/arrow_adapter.h"
+#include "utils/common.h"
+#include "utils/logger.h"
+
+// Typedef for a 64-bit khash table
+KHASH_MAP_INIT_INT64(m64, int64_t)
+
+namespace tiledbsoma {
+
+void IntIndexer::map_locations(const int64_t* keys, int size, int threads) {
+    map_size_ = size;
+    if (size == 0) {
+        return;
+    }
+    if (size < 10) {
+        threads = 1;
+    }
+    if (size < threads)
+        throw std::runtime_error(
+            "The number of keys " + std::to_string(size) +
+            " must be larger than the number of threads " +
+            std::to_string(threads) + " .");
+
+    LOG_DEBUG(fmt::format(
+        "End of Map locations of size {} and {} threads", size, threads));
+
+    LOG_DEBUG(fmt::format(
+        "[Re-indexer] Start of Map locations with {} keys and {} threads",
+        size,
+        threads));
+    hash_ = kh_init(m64);
+    kh_resize(m64, hash_, size * 1.25);
+    LOG_DEBUG(
+        fmt::format("[Re-indexer] Thread pool started and hash table created"));
+    int ret;
```

Comment:
What is `ret` for these `kh_*` functions? They get passed in as an argument but are unused. Is it the return code? Do we need to check that?

Reply:
Yes, that's the return code. In our case we just check the return object.